src/learning/learning.ts

```diff
@@ -0,0 +1,151 @@
+import { InferenceEngine } from '..'
+import { restoreEngine } from '../engines/util'
+import { FastPotential } from '../engines/FastPotential'
+import { PairedObservation, groupDataByObservedValues } from './Observation'
+import { TowerOfDerivatives } from './TowerOfDerivatives'
+import { StepStatus, statusMessage } from './StepResult'
+import { newtonStep } from './newton-step'
+import { localDistributionPotentials } from './objective-functions/util'
+import { objectiveFunction } from './objective-functions/online-learning-objective-function'
+import { SQRTEPS, CUBEROOTEPS } from './vector-utils'
+import { lineSearch } from './line-search'
+
+/** Compute a relative gradient that can be used for detecting if
+ * the initial parameters are a maximizer for the objective function.
+ * @param current - the current (intial) tower of derivatives
+ * For a measure of the relative gradient, we take the maximum reltaive change in
+ * f with respect to the relative change in a parameter.   This is scaled
+ * to avoid excessively small values in the objective function, gradient or
+ * parameters.
+ */
+function relativeGradient (current: TowerOfDerivatives): number {
+  // The relative gradient is the maximum element of the projection of the
+  // gradient onto the parameter vector onto the ascent direction.  To guard
+  // against small values of the ascent direction, we clamp the minimum value
+  // to the square root of epsilon.
+  const numerator = current.gradient.reduce((acc, gs, i) =>
+    Math.max(acc, gs.reduce((acc2, g, jk) => {
+      const representativeX = Math.max(SQRTEPS, Math.abs(current.xs[i][jk]))
+      return Math.abs(g * representativeX)
+    }, 0)), 0)
+  // However the relative gradient can still be effected by the scale of the
+  // objective function.   To fix this, we divide by the greater the current value
+  // of the objective function or a representative value.   For the representative
+  // value we use half of the worst possible value for the log likelihood function.
+  const denominator = Math.max(Math.abs(current.value), 0.5 * Math.abs(Math.log(SQRTEPS)))
+  return numerator / denominator
+}
+
+/** Compute a relative change in the parameters that can be used for detecting if
+ * the maximizer for the objective function has been found.
+ * @param current - the current (intial) tower of derivatives
+ * @param trial - a new set of parameters identified by either a Newton or quasi-
+ *   Newton step.
+ * For a measure of the relative change, we take the maximum relative change in
+ * parameters between the current and trial values.   This is scaled
+ * to avoid excessively small values.
+ */
+function relativeChange (current: TowerOfDerivatives, trial: TowerOfDerivatives): number {
+  return current.xs.reduce((accI, ps, i) =>
+    Math.max(accI, ps.reduce((accJK, p, jk) => Math.max(accJK,
+      Math.abs(p - trial.xs[i][jk]) / Math.max(Math.abs(p), SQRTEPS),
+    ), 0),
+    ), 0)
+}
```

Comment:
nit: formatting
```suggestion
    ), 0),
  ), 0)
}
```

Reply:
The linter is adding this extra indent :<